### ISSUE TITLE
NVENC encoding not available on Windows 7

### ISSUE THREAD

**Issue by ferdnyc** [OpenShot/openshot-qt#3448]
Hi all, a related case, on Win 7, OS 2.5.1, GF 750 Ti - the NVENC doesn't appear at all as a disabled option. I have CUDA 10.1 lib installed if that could be e problem.

![image](https://user-images.githubusercontent.com/23367640/80804289-5e931a00-8bbd-11ea-9d85-e96a8f9c2752.png)

_Originally posted by @Twenkid in https://github.com/OpenShot/openshot-qt/issues/3414#issuecomment-622363935_

**Comment by ferdnyc** [comment]
@Twenkid During the _Preferences_ window loading, a new info appears in the log of OpenShot, there results of HW decoders tests are logged (if any), So, if you see something like:
```
 preferences:WARNING CheckPixel failed testing hardware decoding in preferences (i.e. wrong color found): 3-0
 preferences:WARNING CheckPixel failed testing hardware decoding in preferences (i.e. wrong color found): 3-1
 preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 4-0
 preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 4-1
 preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 7-0
 preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 7-1
```
then no test were success on your system, so you can't choose anything but CPU decoding.

The list of acceleration test/devices numbers in OpenShot's log:
https://github.com/OpenShot/libopenshot/blob/f1e5c9c8a2acfd738406c869d1d4bc512ca41e04/include/Settings.h#L80-L91

Currently NVIDIA decoder is not listed as supported on Windows for OpenShot:
https://github.com/OpenShot/openshot-qt/blob/e0d3776de53dc8b7d105b757e349f0e703f27155/src/windows/preferences.py#L294

As for the Device selection options:
https://github.com/OpenShot/openshot-qt/blob/e0d3776de53dc8b7d105b757e349f0e703f27155/src/windows/preferences.py#L321-L323

-----------------

This were decoders. As for encoders - I don't know where code is.

_Originally posted by @SuslikV in https://github.com/OpenShot/openshot-qt/issues/3414#issuecomment-622939444_

**Comment by ferdnyc** [comment]
One thing you might try is using the relatively-new `openshot-qt-cli.exe` tool (included in the OpenShot application folder since 2.5.0) to run OpenShot from a Windows `command.com` session.

The advantage to that is, you should be able to see any output produced during the encoding process, including console messages that may be printed by the FFmpeg libraries. Those aren't otherwise visible or available in any of the captured logs. Sometimes when there's a processing issue, including hardware encoder failures, it'll dump an error message to the console before falling back to software encoding.

To run OpenShot that way, just open a Command Prompt window, change to the directory where OpenShot is installed (probably `C:\Program Files\OpenShot Video Editor\`), and run `openshot-qt-cli.exe` to launch OpenShot with the console output visible. So:
```cmd
cd "C:\Program Files\OpenShot Video Editor"
openshot-qt-cli.exe
```

(You might have to increase the buffer length in the Command Prompt window first. IIRC the default is only ~ 300 lines? Not enough for noisy OpenShot output. Something like 2000 or 5000 lines is more realistic.)

_Originally posted by @ferdnyc in https://github.com/OpenShot/openshot-qt/issues/3414#issuecomment-622940688_

**Comment by ferdnyc** [comment]
> Currently NVIDIA decoder is not listed as supported on Windows for OpenShot:

Wait, so there *is* some hardware decoding supported on Windows? I thought I read on Reddit that it was not supported at all. (Why I did not bother to report it)

Like @Twenkid I saw the Intel QSV listed, but I thought *that* was a bug. It is showing up on my old laptop too,  which only has an Intel CPU, not an Intel GPU. Should it be seeing an old AMD/ATI GPU? (~10y old, OpenCL 1.3 or something). I thought ATI was not supported either? (Not really sure how to read your link)

On my new machine with both Intel and NVidia GPU are not seen either, except for that single QSV. 
I can not even get it to run the interface on the 'high-end GPU' from the Nvidia settings. 

@ferdnyc 
openshot-qt-cli.exe 
Would it matter to add that one to Nvidia, or is it running other processes in the background? (IRRC, Blender is seen by Nvidia when called from Openshot.

Also, is it documented somewhere yet? What does it do, can you use it to add files on a time/track with a script or something?
A google search gave me this:
https://blueprints.launchpad.net/openshot/+spec/command-line-interface-for-project-rendering

_Originally posted by @MBB232 in https://github.com/OpenShot/openshot-qt/issues/3414#issuecomment-622977716_

**Comment by ferdnyc** [comment]
> @Twenkid During the _Preferences_ window loading, a new info appears in the log of OpenShot, there results of HW decoders tests are logged (if any), So, if you see something like:
> 
> ```
>  preferences:WARNING CheckPixel failed testing hardware decoding in preferences (i.e. wrong color found): 3-0
>  preferences:WARNING CheckPixel failed testing hardware decoding in preferences (i.e. wrong color found): 3-1
>  preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 4-0
>  preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 4-1
>  preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 7-0
>  preferences:WARNING Exception trying to test hardware decoding in preferences (this is expected): 7-1
> ```
> 
> then no test were success on your system, so you can't choose anything but CPU decoding.
> 

Thanks, yes, it reports errors like that.

```
preview_thread:INFO refreshFrame
preview_thread:INFO self.player.Position(): 1
preview_thread:INFO onModeChanged
preview_thread:INFO onModeChanged
preview_thread:INFO previewFrame: 23
preview_thread:INFO self.player.Position(): 23
preview_thread:INFO previewFrame: 32
preview_thread:INFO self.player.Position(): 32
preview_thread:INFO previewFrame: 37
preview_thread:INFO self.player.Position(): 37
preview_thread:INFO previewFrame: 32
preview_thread:INFO self.player.Position(): 32
     ui_util:INFO Initializing UI for dlgPreferences
Hardware decoding device number: 0
Hardware decoding device number: 0
 preferences:WARNING CheckPixel failed testing hardware decoding in preferences
(i.e. wrong color found): 3-0
Hardware decoding device number: 1
Hardware decoding device number: 1
 preferences:WARNING CheckPixel failed testing hardware decoding in preferences
(i.e. wrong color found): 3-1
Hardware decoding device number: 0
Hardware decoding device number: 0
[h264 @ 0000000004d42180] Invalid setup for format dxva2_vld: does not match the
 type of the provided device context.
 preferences:WARNING CheckPixel failed testing hardware decoding in preferences
(i.e. wrong color found): 4-0
Hardware decoding device number: 1
Hardware decoding device number: 1
[h264 @ 0000000004d43880] Invalid setup for format dxva2_vld: does not match the
 type of the provided device context.
 preferences:WARNING CheckPixel failed testing hardware decoding in preferences
(i.e. wrong color found): 4-1
Hardware decoding device number: 0
[AVHWDeviceContext @ 0000000017c8f3c0] Error initializing an MFX session: -3.
Hardware decoding device number: 0
[AVHWDeviceContext @ 0000000017c901c0] Error initializing an MFX session: -3.
 preferences:WARNING Exception trying to test hardware decoding in preferences (
this is expected): 7-0
Hardware decoding device number: 1
[AVHWDeviceContext @ 0000000017c90340] Error initializing an MFX session: -3.
Hardware decoding device number: 1
[AVHWDeviceContext @ 0000000017c90540] Error initializing an MFX session: -3.
 preferences:WARNING Exception trying to test hardware decoding in preferences (
this is expected): 7-1
  OpenType support missing for "Ubuntu", script 11
(...)
DirectWrite: CreateFontFaceFromHDC() failed (Unknown error 0x88985000.) for QFon
tDef(Family="Fixedsys", pointsize=10.5, pixelsize=15, styleHint=5, weight=50, st
retch=100, hintingPreference=0) LOGFONT("Fixedsys", lfWidth=0, lfHeight=-15) dpi
=96

```

_Originally posted by @Twenkid in https://github.com/OpenShot/openshot-qt/issues/3414#issuecomment-622990557_

**Comment by ferdnyc** [comment]
> Wait, so there _is_ some hardware decoding supported on Windows? I thought I read on Reddit that it was not supported at all. (Why I did not bother to report it)

I really have no idea, when it comes to Windows, but it's worth pointing out that hardware _decoding_ isn't really what you want anyway, as it's minimally useful to OpenShot. Where you get the really big bang for your buck is hardware <b>EN</b>coding, which can make a big difference in export speeds to supported formats.

(That "supported formats" can be a very restrictive list or a pretty decent one, it all depends on how many digits were in the price of your GPU (based on US currency or equivalent). Mine is a 2-digit ($50) card, so it manages x264 fine, but little else of use. (...Until MPEG2 makes its big comeback. \#SOON \#NOTREALLY)
 
> Like @Twenkid I saw the Intel QSV listed, but I thought _that_ was a bug. 

Nope, that'd be an available hardware _en_-coding path, which mostly comes down to what encoders FFmpeg reports as being available on the system.

Any of these are still possibilities, though:
1. Exporting with the QSV-accelerated codec would fail when attempted
1. Exporting with the QSV-accelerated codec would use the _CPU_ to "accelerate" the encoding, which if you're lucky would merely be the same as exporting without acceleration. If unlucky, it would actually make things slower
1. Exporting with the QSV-accelerated codec would attempt to use the ATI GPU as the hardware renderer, and fail
1. Exporting with the QSV-accelerated codec would attempt to use the ATI GPU as the hardware renderer, and succeed
1. Your laptop's Intel CPU has an integrated GPU. The ATI chip made it superfluous, and it doesn't have any video output connections and can't be used for screen display. But it's still there and detected by FFmpeg.

That last one I consider more akin to "hardware-detection fanfic" than an actual theory, IOW it's about 0% likely to be accurate. And even if it were somehow correct, a vestigial Intel GPU in a 10-year-old laptop's single-chip integrated CPU package would be laughably underpowered and entirely useless for any sort of accelerated video anything. Especially if it was such a crap GPU that the laptop manufacturer tossed a separate ATI chip onto the motherboard so it could intimidate the sad little i915<sup>1</sup> until it went into hiding.

> It is showing up on my old laptop too, which only has an Intel CPU, not an Intel GPU. 

> Should it be seeing an old AMD/ATI GPU? (~10y old, OpenCL 1.3 or something). I thought ATI was not supported either? (Not really sure how to read your link)

It helps to have the secret decoder ring for those numbers, which is:
https://github.com/OpenShot/openshot-qt/blob/e0d3776de53dc8b7d105b757e349f0e703f27155/src/windows/preferences.py#L337-L352

So on Windows, detection is run for 0, 3, 4, and 7 which means Direct3D 9, Direct3D 11, and Intel QSV, in addition to no-acceleration (CPU-only).

Now, in theory Direct3D 9 and 11 should offer at least the possibility of driving _any_ GPU regardless of the manufacturer, under Windows. Though I have no idea what actually works in practice. There's some documentation on the hardware encoding state here, though it may be out of date: https://github.com/OpenShot/libopenshot/blob/develop/doc/HW-ACCEL.md

(If it's _not_, it indeed says that everything is broken on Windows, and Intel QSV is broken **everywhere**.)

> On my new machine with both Intel and NVidia GPU are not seen either, except for that single QSV.
> I can not even get it to run the interface on the 'high-end GPU' from the Nvidia settings.

Yeah, I'm betting we don't support secondary GPUs, since using those would require making the necessary API calls to activate the discrete chip, and as you say OpenShot doesn't even know how to do that for _itself_, never mind FFmpeg.

> @ferdnyc
> openshot-qt-cli.exe
> Would it matter to add that one to Nvidia, or is it running other processes in the background? (IRRC, Blender is seen by Nvidia when called from Openshot.

Other way around, I think — Blender is able to see and make use of the Nvidia GPU, because it knows how. All we do (when Blender GPU is enabled in the OpenShot preferences) is add some code to our Blender script that says we'd like it to please go ahead and GPU its brains out, if at all possible. The actual "how" of making that happen is part of Blender itself, where fortunately hardware acceleration has long been a priority. (For obvious reasons: 3D rendering _sloooow_.)

So, no, wouldn't make any difference on the OpenShot/FFmpeg side.

> Also, is it documented somewhere yet? What does it do, can you use it to add files on a time/track with a script or something?

LOL, why does [everyone immediately go there](https://github.com/OpenShot/openshot-qt/pull/3106#issuecomment-559986487)??!? Nope, the CLI launcher is just that, a _launcher_ for the same GUI interface as regular OpenShot, with no changes to any part of the user interface or functionality. It exists for one reason and one reason only:

On Linux and MacOS, when you run OpenShot from a terminal window you get log messages from the application spewed out real time, something that's not possible using `openshot-qt.exe` even if you run it from a Windows Command Prompt or PowerShell. And while _most_ of the output ends up in a logfile and can just be reviewed there, not only is it sometimes just more convenient to have an immediate stream of messages, but there are certain messages that _**only**_ appear on the console. Primarily, messages printed directly to the console by the FFmpeg libraries, which don't get captured to any log. So, `openshot-qt-cli.exe` simply allows OpenShot to be run on Windows the same way it's always been on the other OSes.

Which is why, I don't think it should be documented anywhere. I think it's best to keep it as a "hidden" feature, present in the install for use by those who know (or are told) about it, and discoverable by anyone curious enough to go poking around, but otherwise not officially acknowledged at all.

> A google search gave me this:
> https://blueprints.launchpad.net/openshot/+spec/command-line-interface-for-project-rendering

Totally unrelated, in fact I never knew about that until today. It's also dated 10 years ago, meaning it was back during the OpenShot 1.0 days. OpenShot 2.0 was a complete rewrite, all new backend libraries (MLT out, FFmpeg-based libopenshot in) and frontend toolkit (Qt via PyQt5, replacing PyGTK2) so anything dated from before 2015 (maybe mid-2014) would be previous-generation stuff and therefore no longer relevant/applicable.

#3106 is the PR that incorporated `openshot-qt-cli.exe` into the Windows packaging. It was basically a random, pseudo-spontaneous idea not really born of any prior discussions or requests in particular. (In _general_ there must've been a half-dozen or more times it seemed like it would've been a handy thing to have, for some reason or other, and eventually that thought reached critical mass.)

#### Notes
1. (I'm guessing. The laptop's not nearly old enough to have an `i810` in it, plus I don't think those existed as GPU modules in integrated single-chip mobile processors, only as standalone GPU chips. Pretty sure the i915 was used in single-chip, though. And then came the i965, which is primarily — (exclusively?) — found as part of a laptop CPU package.)

**Comment by MBB232** [comment]
> > Would it matter to add that one to Nvidia, or is it running other processes in the background? (IRRC, Blender is seen by Nvidia when called from Openshot.
> 
> Other way around, I think — Blender is able to see and make use of the Nvidia GPU, because it knows how. All we do (when Blender GPU is enabled in the OpenShot preferences) is add some code to our Blender script that says we'd like it to please go ahead and GPU its brains out, if at all possible. The actual "how" of making that happen is part of Blender itself, where fortunately hardware acceleration has long been a priority. (For obvious reasons: 3D rendering _sloooow_.)
> 
> So, no, wouldn't make any difference on the OpenShot/FFmpeg side.

Perhaps, but the NVIDIA software (on Windows) seems to be pretty involved in the switching.  For instance, LibreOffice and Gimp normally run without the GPU, but if OpenCL gets checked it automatically switches to the high-end NVidia. 
(Remind me to test if unchecking that box you talk about matters for Openshot. )

There are also profiles where some programs are automatically recognized and you can add others manually. However, VLC and MPCclassic get rejected, so there may be some problems with FFMpeg. (Or some commercial competitor is paying NVidia some big bugs to reject them). 
However, renaming the exe to a different allows them to be assigned a profile. (Something that fails for Openshot because if it gets renamed it fails to run). 

Is the FFmpeg of openshot a separate exe that I could try to add, or is it completely internal?

How widely is libopenshot used, is it in Blender too?
Perhaps you can get some NVIDIA people to take a look at it, they did some awesome AI features for Blender rendering recently
https://code.blender.org/2019/07/accelerating-cycles-using-nvidia-rtx/
https://www.google.com/search?ei=NNGwXpTeHoKxkwXk9a7YAg&q=nvidia+blender+ai+denoiser&oq=nvidia+blender+AI&gs_lcp=CgZwc3ktYWIQAxgAMggIIRAWEB0QHjoECAAQRzoCCAA6BggAEBYQHlCgMljAYGCCamgAcAJ4AIABTogB3wGSAQEzmAEAoAEBqgEHZ3dzLXdpeg&sclient=psy-ab


> 
> > Also, is it documented somewhere yet? What does it do, can you use it to add files on a time/track with a script or something?
> 
> LOL, why does [everyone immediately go there](https://github.com/OpenShot/openshot-qt/pull/3106#issuecomment-559986487)??!? Nope, the CLI launcher is just that, a _launcher_ for the same GUI interface as regular OpenShot, with no changes to any part of the user interface or functionality. It exists for one reason and one reason only:

Hey, I'm a _Windows_ user. The _only_ reason I can imagine people _voluntarily_ using a command line is to run scripts 😁, especially for a graphic program like a video editor. (Even when I had to use the last of DOS, it already had some text-based form of a GUI )

I was already imagining a script that could convert subtitle files to SVG, that could then be loaded into Openshot at the right time/track. 
Not that I really understand _why_ people would want to use Openshot for subtitles, but it seems to be a popular use on Youtube tutorials and Reddit questions.

**Comment by MBB232** [comment]
>So on Windows, detection is run for 0, 3, 4, and 7 which means Direct3D 9, Direct3D 11, and Intel QSV, in addition to no-acceleration (CPU-only). Now, in theory Direct3D 9 and 11 should offer at least the possibility of driving _any_ GPU regardless of the manufacturer, what with Windows being the 900-pound gorilla able to bend hardware manufacturers to its API whims. I have no idea what actually might _work_ in practice, though. There's some documentation on the hardware encoding state here, though it may be out of date: https://github.com/OpenShot/libopenshot/blob/develop/doc/HW-ACCEL.md
> 
> (If it's _not_ out of date, then it indeed says that everything is broken on Windows, and Intel QSV is broken **everywhere**.)
> 
> > On my new machine with both Intel and NVidia GPU are not seen either, except for that single QSV.
> > I can not even get it to run the interface on the 'high-end GPU' from the Nvidia settings.
> 
> Yeah, I'm betting we don't support secondary GPUs, since using those would require making the necessary API calls to activate the discrete chip, and as you say OpenShot doesn't even know how to do that for _itself_, never mind FFmpeg.

Thanks for translating, I don't think I would have figured out that keyring correctly on my own.
I'll take another look, perhaps I remember it wrong and it does show some encoders, but I think the 'everything is broken' applies to windows, considering the high number of issues about hardware acceleration I have seen about it. 

Also good table you linked - I think I saw it before but could not find it. 
I do have CUDA installed, so NVENC should work (perhaps it wants CUDA 9 iso 10?)
It does not list OpenCL and Vulcan -would those not be useful for cross-platform video editing?

It does however explicitly state that multi-GPU should not be an issue: 
But only on Linux you can select separate cards for encoding vs decoding, on Windows you have to use the sane card for both:

>If the computer has multiple graphics cards installed, you can choose which should be used by libopenshot. Also, you can optionally use one card for decoding and the other for encoding (if both cards support acceleration). This is currently only supported on Linux, due to the device name FFmpeg expects (i.e. /dev/dri/render128). Contributions welcome if anyone can determine what string format to pass for Windows and Mac.

**Comment by MBB232** [comment]
Some sites I found that may be of use in case people want to improve this in the future: 
FFMPEG https://trac.ffmpeg.org/wiki/HWAccelIntro
NVIDIA https://developer.nvidia.com/ffmpeg

GPU Accelerated Computing with Python
 https://developer.nvidia.com/how-to-cuda-python?utm_content=bufferbfedb&utm_medium=social&utm_source=facebook.com&utm_campaign=buffer
https://devblogs.nvidia.com/numba-python-cuda-acceleration/

**Comment by Twenkid** [comment]
Re QSV - it works and is pretty good to have, especially on laptops. I think I first discovered it in 
 a daily build in the autumn.

However in order to have QSV operating, if you had a discrete GPU as well, you may have to **turn the onboard graphics explicitly from the BIOS**, because the "Auto" mode could be the BIOS to turn it off on boot, when it detects another GPU. In the latter case the QSV encoding is available in the menu (in my case), but it won't work when you try to export.

**Comment by ferdnyc** [comment]
> I do have CUDA installed, so NVENC should work (perhaps it wants CUDA 9 iso 10?)

NVENC doesn't even require CUDA except to _build_ it, CUDA is a development platform more than a service, software built with GPU support should be pretty much self-contained. But I don't know whether Windows FFmpeg supports acceleration there the same way it does on Linux, and even if it does what's included in the build we package.

> It does not list OpenCL and Vulcan -would those not be useful for cross-platform video editing?

RadeonOpenCompute, which is basically an OpenCL implementation AIUI, is mentioned in passing as a possible way to accelerate effects. Which is about right -- GPU computing might be good for some intensive computations, including the image processing operations performed by some effects and filters, but video encoding doesn't necessarily benefit from it that much. At least, not _automatically_. 

Blender benefits from CUDA/OpenCL/etc. Hugely. Possibly some of our effect algorithms could. (They'd have to be rewritten for the GPU, though, just like Blender's rendering algorithms have been.)

That's why things like NVENC exist. It uses the GPU to run highly-optimized algorithms exclusively designed to accelerate video encoding (and only with certain specific codecs). But GPU computing is only really useful if you also have code specifically designed to take advantage of it, which we don't. Even the fastest, beefiest GPU isn't a free speedup for general-purpose software. You can't run applications on a GPU. They're so much faster for certain tasks, because they can ONLY do certain tasks... and only when those tasks are crafted so they lean in to the GPU's strengths.

You don't get GPU computing acceleration for free on the hardware side either. MOST consumer video cards don't have the kind of power to accelerate CUDA algorithms significantly. So the user base for this kind of general-purpose GPU computing is relatively small even compared to the OpenShot user base as a whole.

My video card is mostly meant for playback, and it accelerates decoding fine, plus gives me a reasonable boost for video encoding. I think last I checked it was something like 2-4x as fast as my CPU when encoding x264.

But for GPU computing? Like I've said, it's SLOWER than my CPU is, when rendering in Blender. There is no real benefit to CUDA & friends for anyone with a system like mine, one that wasn't built for power and uses a non-"gamer"-level, sub-$200 graphics processor. So, while there might be speedups to be had for some users, I wouldn't want to leave the remaining majority out in the cold by focusing on features exclusively targeted towards users with high-end hardware. (Or, even worse, making OpenShot **only** support those systems.)

> It does however explicitly state that multi-GPU should not be an issue: 

Yeah, that's questionable from my POV. That statement applies to Linux, and it's _assumed_ that the same thing can be done on Windows. I'm not as optimistic on that point. Especially when we're talking about FFmpeg, a set of libraries that aren't really native to Windows. They're effectively cross-compiled Linux libs. That there could be working hardware acceleration **at all** amazes me. I don't think we can make any assumptions.

The other thing is, that's talking about multiple SEPARATE GPUs, which is a common thing in dedicated video editing rigs, rendering farms, and crypto currency mining. You take a server motherboard, stick four $1000 GPUs in it, and try to keep it from catching on fire while it churns through whatever you throw at it.

In those setups, you need to be able to specify which of the cards you run a job on, and you can do that by telling FFmpeg which `/dev/dri/` device to use. But laptop GPU switching are a completely different thing, it doesn't work the same way at all.

You also mentioned Vulkan, which I can't offer any useful info on. I know nothing about it beyond the fact that it exists.

**Comment by ferdnyc** [comment]
> You take a server motherboard, stick four $1000 GPUs in it, and try to keep it from catching on fire 

(I meant that literally, BTW.)

**Comment by easy-and-simple** [comment]
Your openshot is real openshit like all of you dumb and useless morons
If you have not programming skills then don't do your openshit but go to clean closets
Handicapped lamers that can't do simple thing are telling others that problem is in their too good hardware or too bad hardware and that all hardware is not compatible with their brainless garbage code
After I installed normal and not restricted copt of ffmpeg in export option have appeared nvencode option but it is not working because these insolent and dumb morons can't even pass correct parameters to ffmpeg
This useless insolent moron SuslikV told me that their openshit is open source so I should program features that they have declared like their fake HW support or to use other software
Codecs included in openshit are half size because they are not build with HW support
![bad-codecs](https://user-images.githubusercontent.com/64713495/81197523-10b95000-8fc9-11ea-8e2a-aede5070d7c5.JPG)

Forget this real openshit because it is handled by handicapped lamers that are too dumb useless and insolent to do something

**Comment by SuslikV** [comment]
because I was mentioned here, @easy-and-simple comes from the https://github.com/OpenShot/openshot-qt/issues/3439 .

**Comment by ferdnyc** [comment]
@easy-and-simple Thanks for your input!

My stupidity goes even deeper, if it were up to me we wouldn't support Windows at all.

**Comment by MBB232** [comment]
> @easy-and-simple Thanks for your input!
> 
> My stupidity goes even deeper, if it were up to me we wouldn't support Windows at all.

Not sure what that says about _my_ stupidity, wanting to support a project that does not really want me. 🤦‍♂️
 (Though if most windows users are as rude, I'm suddenly a lot more understanding of your position.) 
Lets hope that the Windows Subsystem for Linux gets hardware support soon, you can run the backend of Openshot on that and I get the best of both worlds. 

>You also mentioned Vulkan, which I can't offer any useful info on. I know nothing about it beyond the fact that it exists.

I don't know much either, but I thought it was supposed to be the modern version of OpenGL for crossplatform video drivers. So expected it to feature pretty heavy in the FOSS video community. 

Also I think the new Intel Xe videocard is supposed to be depending on it. So if Openshot were to ask them, they may be willing to offer a bunch of resources (develloppers/money/hardware?) if you offered them to support it? 

https://www.phoronix.com/scan.php?page=news_item&px=Intel-ANV-Preps-Multi-GPU-Xe
https://www.tomshardware.com/news/intel-xe-graphics-all-we-know
[For developpers ](https://www.engadget.com/2020/01/09/intel-xe-d1g-software-development-vehicle/?guccounter=1&guce_referrer=aHR0cHM6Ly9kZXZlbG9wZXJzLnNsYXNoZG90Lm9yZy9zdG9yeS8yMC8wMS8wOS8xNzQ0MjQzL2ludGVscy1maXJzdC1kaXNjcmV0ZS1ncHUtaXMtYnVpbHQtZm9yLWRldmVsb3BlcnM&guce_referrer_sig=AQAAAIFg9vNX-0WcT8EVR_mjZxKSK2VyxeLkbzv7T_ibdbAhwG_CzzC0A60RlbtADZI9GIQo9OPfYHNU3N0-OWKn9d4xlDKkBYh1m-fkeBMHlRbw0ql5qgMLzSgWlnYXtrWs-9BBKCmJomrVdC1G6JFWD-0WRxe_pX-usF7BUDF2SeQJ#comments)

Then, they are also talking about OneAPI, though that may be a CUDA alternative. They claim to have a cloudbased programming sandbox
https://software.intel.com/content/www/us/en/develop/tools/oneapi.html

**Comment by MBB232** [comment]
> Other way around, I think — Blender is able to see and make use of the Nvidia GPU, because it knows how. All we do (when Blender GPU is enabled in the OpenShot preferences) is add some code to our Blender script that says we'd like it to please go ahead and GPU its brains out, if at all possible. The actual "how" of making that happen is part of Blender itself, where fortunately hardware acceleration has long been a priority. (For obvious reasons: 3D rendering _sloooow_.)

Actually, no it is not: the GPU is only used when NOT selected in Openshot. (Maybe when I do select it, it may want to force it on the Intel GPU instead of Blender automatically picking NVidia)
As can be seen in the screenshot in the 3D graph, the run before the peek is with 'Use Blender GPU render' Disabled, and the run after is with it enabled. (though it still claims a use of 5%?)
It was also a lot faster on the first run.

As for QSV, I seem to be one of the rare lucky ones for who it is working out of the box. (At least, it exports without errors and shows activity on the Intel Video encoder graph, though it does not seem to be faster) 
But only if run the normal OpenShot exe, it fails if I run the openshotCLI exe

**Comment by MBB232** [comment]
@MBB232 wrote: 
> Lets hope that the Windows Subsystem for Linux gets hardware support soon, you can run the backend of Openshot on that and I get the best of both worlds.

Speak of the devil: https://devblogs.microsoft.com/directx/directx-heart-linux/

> **DirectX is coming to the Windows Subsystem for Linux**
...
If the host has multiple GPUs, all GPUs are projected and available to the Linux environment (assuming all of these GPUs are running WDDMv2.9 drivers).
...
Applications running inside of the Linux environment have the same access to the GPU as native applications on Windows.
...
The D3D12 API can be used for offscreen rendering and compute, but there is no swapchain support to copy pixels directly to the screen (yet 😊).
OpenGL, OpenCL & Vulkan
..We have recently announced work on mapping layers that will bring hardware acceleration for OpenCL and OpenGL on top of DX12. We will be using these layers to provide hardware accelerated OpenGL and OpenCL to WSL through the Mesa library...
...
We are pleased to announce that NVIDIA CUDA acceleration is also coming to WSL! CUDA is a cross-platform API and can communicate with the GPU through either the WDDM GPU abstraction on Windows or the NVIDIA GPU abstraction on Linux..
...
GUI Applications
At //build we announced that support for Linux GUI applications is coming to WSL. While today WSL is a console only experience, soon you’ll be able to use your favorite Linux IDE or other GUI application alongside your other Windows applications on your Windows desktop.


Even GUI support is coming, so no even need  to implement https://github.com/OpenShot/openshot-qt/issues/3102 first. (though it may make it smoother)

In fact, the whole DirectX api might some day come to Linux; so even more possible combinations that can start giving bugs :-)
